crypto/fipsmodule/sha/sha3.c

```diff
@@ -236,7 +257,62 @@ int SHAKE_Init(KECCAK1600_CTX *ctx, size_t block_size) {
   return 0;
 }
 
+int SHAKE_Absorb(KECCAK1600_CTX *ctx, const void *data, size_t len) {
+  if (ctx == NULL) {
+    return 0;
+  }
+
+  if (len == 0) {
+    return 1;
+  }
+
+  return FIPS202_Update(ctx, data, len);
+}
+
+// SHAKE_Final is a single-shot API and can be called once to finalize absorb and squeeze phases
+// |ctx->state| restricts consecutive calls to FIPS202_Finalize
+// Function SHAKE_Squeeze should be used for incremental XOF output
 int SHAKE_Final(uint8_t *md, KECCAK1600_CTX *ctx, size_t len) {
   ctx->md_size = len;
-  return SHA3_Final(md, ctx);
+  if (ctx->md_size == 0) {
+    return 1;
+  }
+
+  if (FIPS202_Finalize(md, ctx) == 0) {
+    return 0;
+  }
+
+  Keccak1600_Squeeze(ctx->A, md, ctx->md_size, ctx->block_size, ctx->state);
+  ctx->state = KECCAK1600_STATE_FINAL;
+
+  FIPS_service_indicator_update_state();
+
+  return 1;
+}
+
+// SHAKE_Squeeze can be called multiple times
+// SHAKE_Squeeze should be called for incremental XOF output
+int SHAKE_Squeeze(uint8_t *md, KECCAK1600_CTX *ctx, size_t len) {
+  ctx->md_size = len;
+
+  if (ctx->md_size == 0) {
+    return 1;
+  }
+
+  if (ctx->state == KECCAK1600_STATE_FINAL) {
+    return 0;
+  }
+
+  if (ctx->state == KECCAK1600_STATE_ABSORB) {
+    // Skip FIPS202_Finalize if the input has been padded and the last block has been processed
```

Comment:
Maybe move this comment one line up because it's talking about skipping this if statement (if I understood correctly).